### PR TITLE
Added Flow types to pause actions and reducer

### PR DESCRIPTION
**Pull request by montogeek**
Associated Issue: #1004 

### Summary of Changes

Added Flow types to pause actions
Added Flow types to pause reducers

### Testing

[x] passes `npm test`
[x] passes `npm run lint`

Notes:
Flow version: 0.35.0, I used the last one since it was not specified in `.flowconfig`, although 0.34.0 is specified in `package.json`, it took the chance to use the last version.

- There are errors related with the `fromJS` utility, since it is not Flow typed.
- Flow 0.35.0 started complaining about these lines https://github.com/devtools-html/debugger.html/blob/master/public/js/utils/utils.js#L207 and https://github.com/devtools-html/debugger.html/blob/master/packages/devtools-local-toolbox/public/js/utils/utils.js#L163 with the error: `call of method assign: Expected object instead of T`
- `pause` reducer (https://github.com/devtools-html/debugger.html/blob/master/public/js/reducers/pause.js#L41) is handling an `source` action (https://github.com/devtools-html/debugger.html/blob/master/public/js/actions/sources.js#L195)
- This constant/action does not longer exists: https://github.com/devtools-html/debugger.html/blob/master/public/js/reducers/pause.js#L56
- Unfortunately Flow does not support object computed keys in Object type declarations, thats why this is hardcoded: https://github.com/devtools-html/debugger.html/compare/master...montogeek:flow-type-pause-actions-reducer?expand=1#diff-195a73de407881358c8dfc0c5891f7feR114

### CODE REVIEW

**Comment by jasonLaster** [comment]
This looks great @montogeek! Especially PauseState and ThunkArgs. I'll take a look monday morning. 

At first glance it looks like both you and @Zacqary added a frame [type](https://github.com/devtools-html/debugger.html/pull/1121)

**Comment by jasonLaster** [comment]
@montogeek just rebased your PR

File: src/types.js [hunk 1]
```diff
@@ -6,6 +6,34 @@
...
+}
+
+export type Pause = {
+  frames: Frame[],
+  why: Why
+}
+
+export type Expression = {
+  id: number,
+  input: any
```
Comment (wldcordeiro) [inline src/types.js]: Is the type of `input` really `any`? Or could we use one that is a bit more specific?

File: src/actions/pause.js [hunk 2]
```diff
@@ -177,8 +183,8 @@ function selectFrame(frame) {
  * @memberof actions/pause
  * @static
  */
-function loadObjectProperties(grip) {
-  return ({ dispatch, client }) => {
+function loadObjectProperties(grip: any) {
```
Comment (wldcordeiro) [inline src/actions/pause.js]: Same question here, do we need to use `any` or can we be more specific?

**Comment by jasonLaster** [comment]
Going through this now. There's some really good stuff here and I can see how the flow score went up!

tracking my thoughts as i go through it here

- **?** ---  There are errors related with the `fromJS` utility, since it is not Flow typed.

- **fixed** https://github.com/devtools-html/debugger.html/pull/1199 ---  Flow 0.35.0 started complaining about these lines https://github.com/devtools-html/debugger.html/blob/master/public/js/utils/utils.js#L207 and https://github.com/devtools-html/debugger.html/blob/master/packages/devtools-local-toolbox/public/js/utils/utils.js#L163 with the error: `call of method assign: Expected object instead of T`


- **?** -- `pause` reducer (https://github.com/devtools-html/debugger.html/blob/master/public/js/reducers/pause.js#L41) is handling an `source` action (https://github.com/devtools-html/debugger.html/blob/master/public/js/actions/sources.js#L195)


- **removed** ---  I'll verify that we don't need it in a bit [blame](https://github.com/montogeek/debugger.html/blame/flow-type-pause-actions-reducer/src/reducers/pause.js#L71-L77) --- This constant/action does not longer exists: https://github.com/devtools-html/debugger.html/blob/master/public/js/reducers/pause.js#L56

- **?** -- Unfortunately Flow does not support object computed keys in Object type declarations, thats why this is hardcoded: https://github.com/devtools-html/debugger.html/compare/master...montogeek:flow-type-pause-actions-reducer?expand=1#diff-195a73de407881358c8dfc0c5891f7feR114

File: src/types.js (hunk 1 above)
Comment (montogeek) [inline src/types.js]: We can use one more specific :)

**Comment by montogeek** [comment]
Thanks for the thoughtful review, I will continue working on it.

About the hardcoded string, it is this: `@@dispatch/promise`, it the name of the promise middleware https://github.com/devtools-html/debugger.html/blob/master/src/utils/redux/middleware/promise.js#L9

I tried to import it and use it as key, but it is not supported by Flow:
```
  ...
  [PROMISE]: T
```

**Comment by jasonLaster** [comment]
> re: `hardcoded string`

@montogeek that's an interesting question. We should be handling it in Sources and breakpoints which is typed. @jlongster do you know the answer?

File: src/utils/pause.js [hunk 3]
```diff
@@ -1,7 +1,10 @@
+// @flow
 const { Frame } = require("../tcomb-types");
 const { getOriginalLocation } = require("./source-map");
 
-function updateFrameLocations(frames) {
+import type { Frame as TFrame } from "../types";
+
+function updateFrameLocations(frames: TFrame[]): Promise<TFrame[]> {
```
Comment (jasonLaster) [inline src/utils/pause.js]: small nit, when there's a conflict we use `FrameType`

https://github.com/montogeek/debugger.html/blob/flow-type-pause-actions-reducer/src/actions/sources.js#L273-L275

**Comment by montogeek** [comment]
@jasonLaster Thank you so much! Sorry for not catching those details!

**Comment by jasonLaster** [comment]
Impossible to catch everything!

I double and triple checked that too

**Comment by montogeek** [comment]
YAY! 💃

**Pull request merged**